### ISSUE TITLE
Several issues I found

### ISSUE THREAD

**Issue by xingminw** [plgreenLIRU/Gaussian-Process-SparseGP-Tutorial#3]
Thank you for providing this tutorial which is helpful for people that are new to this topic. 
I know this is more like a tutorial code, however, I found several parts that could be misleading:

* After you find the inducing variables, you essentially use the inducing points as a subset to perform a GP without using all the points, which is not correct;
* When you calculate the negative likelihood, you did not use a good way to simplify the matrix computation. One major advantage of sparse variational GP is to reduce the complexity from O(n3) to O(m2n). However, the way you implement it is still ordered by n3 since you directly take the inverse of the `matrix c`. Besides, the `slogdet` operator can also be simplified. 

For your information, here are two matrix operations that are essentially very slow:

![image](https://user-images.githubusercontent.com/44205791/230425934-59ae924f-0264-4aef-8863-9a917bf28143.png)

For the matrix inversion, for example, the inversion at least can be speeded up significantly through the matrix inversion lemma: 
$$(QQ^T+\sigma^2I)^{-1}=\sigma^{-2}[I-Q(\sigma^2I+Q^TQ)^{-1}Q^T]$$
which reduces O(n3) to O(nm2).
Besides, the `logdet` can also be simplified using:
$$det(I_n+QQ^T)=det(I_m+Q^TQ)$$

Just as I mentioned, I know this is a tutorial code. However, I do believe it would be better to correct these issues so that the algorithm runs with the correct order O(m2n).

**Comment by xingminw** [comment]
To fix the later issue, here is my implementation (I use different names it is slightly different but should be easy to understand):

```python
 # inv matrix, apply matrix inversion lemma: (QQ.T+s^2*I)^-1 = s^-2[I-*Q(s^2*I+Q.T Q)^-1Q.T]
inv_matrix_left = np.linalg.cholesky(inv_matrix_k_uu)
matrix_q_left = matrix_k_fu @ inv_matrix_left
matrix_q_smaller = matrix_q_left.T @ matrix_q_left
q_right_inv_mat = np.linalg.inv(sigma ** 2 * np.eye(current_inducing_num) + matrix_q_smaller)
inv_matrix_c = np.eye(self.points_num) - matrix_q_left @ q_right_inv_mat @ matrix_q_left.T
inv_matrix_c /= sigma ** 2

# dete matrix, det(Im + AB) = det(In + BA) and det(aI_m) = a^m
log_det_diff = np.log(1 / sigma) * 2 * self.points_num
equiv_matrix_c = matrix_q_smaller / (sigma ** 2) + np.eye(current_inducing_num)
sign, log_det_c = np.linalg.slogdet(equiv_matrix_c)
log_det_c = log_det_c * sign - log_det_diff

# # slow method to perform inverse (direct inversion, very slow)
# matrix_c = matrix_q_ff + np.eye(self.points_num) * sigma ** 2
# inv_matrix_c = np.linalg.inv(matrix_c)

# # slow method to calculate the determinant (direct calculating logdet, very slow)
# sign, log_det_c = np.linalg.slogdet(matrix_c)
# log_det_c = sign * log_det_c
```

**Comment by plgreenLIRU** [comment]
Hello, thank you for your comment 🙂

Sorry I've been away on holiday and am just picking stuff up again; just to let you know that I've seen you're comment and and should get to it soon. To be honest I'd forgotten that I'd even made the tutorial!

**Comment by plgreenLIRU** [comment]
> To fix the latter issue, here is my implementation (I use different names it is slightly different but should be easy to understand):
> 
> ```python
>  # inv matrix, apply matrix inversion lemma: (QQ.T+s^2*I)^-1 = s^-2[I-*Q(s^2*I+Q.T Q)^-1Q.T]
> inv_matrix_left = np.linalg.cholesky(inv_matrix_k_uu)
> matrix_q_left = matrix_k_fu @ inv_matrix_left
> matrix_q_smaller = matrix_q_left.T @ matrix_q_left
> q_right_inv_mat = np.linalg.inv(sigma ** 2 * np.eye(current_inducing_num) + matrix_q_smaller)
> inv_matrix_c = np.eye(self.points_num) - matrix_q_left @ q_right_inv_mat @ matrix_q_left.T
> inv_matrix_c /= sigma ** 2
> 
> # dete matrix, det(Im + AB) = det(In + BA) and det(aI_m) = a^m
> log_det_diff = np.log(1 / sigma) * 2 * self.points_num
> equiv_matrix_c = matrix_q_smaller / (sigma ** 2) + np.eye(current_inducing_num)
> sign, log_det_c = np.linalg.slogdet(equiv_matrix_c)
> log_det_c = log_det_c * sign - log_det_diff
> 
> # # slow method to perform inverse (direct inversion, very slow)
> # matrix_c = matrix_q_ff + np.eye(self.points_num) * sigma ** 2
> # inv_matrix_c = np.linalg.inv(matrix_c)
> 
> # # slow method to calculate the determinant (direct calculating logdet, very slow)
> # sign, log_det_c = np.linalg.slogdet(matrix_c)
> # log_det_c = sign * log_det_c
> ```

OK good stuff. So to check I'll start with the calculation of the matrix inverse. In general we've got the matrix inverse lemma 

$$( \sigma^2 + AA^T)^{-1} = \sigma^{-2}( I_n - A (\sigma^2I_m + A^TA)^{-1} A^T )$$

where $A$ is $n \times m$. We compute $K_{mm}^{-1}$ then take a Cholesky decomp so that $K_{mm}^{-1}=QQ^T$. Ultimately then, we want to find the inverse of

$$I_n \sigma^2 + K_{nm}K_{mm}^{-1}K_{mn}$$

$$= I_n \sigma^2 + K_{nm}QQ^TK_{mn}$$

$$= I_n \sigma^2 + (K_{nm}Q)(K_{nm}Q)^T$$
and so now we can use the matrix inverse lemma to find that

$$ (I_n \sigma^2 + K_{nm}K_{mm}^{-1}K_{mn})^{-1} = \sigma^{-2} (I_n - K_{nm}Q (\sigma^2I_m + Q^T K_{mn}K_{nm} Q)^{-1} Q^TK_{mn}) $$

Do you agree?

**Comment by xingminw** [comment]
Yes, another issue is what I mentioned at the very beginning. After you find the inducing variables, it seems that you only utilize the inducing variables to calculate the predictive distribution. This might not be correct.

**Comment by plgreenLIRU** [comment]
> Yes, another issue is what I mentioned at the very beginning. After you find the inducing variables, it seems that you only utilize the inducing variables to calculate the predictive distribution. This might not be correct.

Ah right yeah. OK so with a standard GP our predictive distribution is calculated by evaluating 

$$  p(f_* | \boldsymbol{y}) = \int p(f_*|\boldsymbol{f})p(\boldsymbol{f}|\boldsymbol{y}) df $$

but now we have

$$  p(f_* | \boldsymbol{y}) = \int p(f_* | \boldsymbol{f}, \boldsymbol{f}_m)p(\boldsymbol{f}, \boldsymbol{f}_m | \boldsymbol{y})
    d \boldsymbol{f}d\boldsymbol{f}_m $$

where, after a bit of maths, we get to

$$     p(f_* | \boldsymbol{y}) \approx 
    \int \int
    \phi(\boldsymbol{f}_m)
    p(\boldsymbol{f}_m | \boldsymbol{y}) d \boldsymbol{f}_m     $$

If we constrain $\phi$ to be Gaussian with mean $\mu$ and covariance matrix $\boldsymbol{A}$ then we end up with

$$
p(f_* | \boldsymbol{y})
\approx \mathcal{N}(f_*; \boldsymbol{k}_{\*}^T \boldsymbol{K}_{mm}^{-1} \boldsymbol{\mu}, k_{\*\*} - \boldsymbol{k}_{\*}^T \boldsymbol{K}_{mm}^{-1} \boldsymbol{k}_{\*} + \boldsymbol{k}_{\*}^T \boldsymbol{B}\boldsymbol{k}_{\*}) $$

where 

$$ \boldsymbol{B} = \boldsymbol{K}_{mm}^{-1}\boldsymbol{A}\boldsymbol{K}_{mm}^{-1} $$

(and the optimal $\mu$ and $A$ are given in the paper). Does that sound OK to you? 

Oh and would you be happy to review a Pull Request once I make the changes?

**Comment by xingminw** [comment]
It looks good. 

Sure I would like to do that, I might not go into the very details and I can pull your latest change to have a quick check.
